### PR TITLE
Search sagas

### LINKED ISSUE #2871: Create Search Saga code
To make #2737 easier to review I've separated out the saga code itself so the patches are smaller. This issue is just to add the saga support to the app.

### PR DESCRIPTION
**Pull request by tofumatt**
Adds the sagas and tests for them in preparation of #2737. Doesn't use them yet. 😄 

I separated this bit out so things are easier to review.

Fixes #2871.

### CODE REVIEW

File: tests/unit/core/sagas/testUtils.js [hunk 1]
```diff
@@ -6,13 +6,23 @@ import SagaTester from 'redux-saga-tester';
...
 
 
 describe('Saga utils', () => {
+  it('does not allow usage of dispatch from a saga', () => {
+    const fakeLog = { error: sinon.stub() };
+    const errorHandler = createErrorHandler('some-error-handler', {
+      log: fakeLog });
+    errorHandler.dispatch('ANYTHING');
+
+    sinon.assert.calledWith(fakeLog.error,
```
Comment (tofumatt) [inline tests/unit/core/sagas/testUtils.js]: This was lacking test coverage so I added it.

File: src/amo/sagas/index.js [hunk 2]
```diff
@@ -12,10 +12,10 @@ import addons from 'core/sagas/addons';
 
 // Export all sagas for this app so runSaga can consume them.
 export default function* rootSaga() {
-  yield [
+  yield all([
```
Comment (kumar303) [inline src/amo/sagas/index.js]: Was this necessary? I think `fork()` already instructs the middleware to make a non-blocking call.

File: src/core/actions/search.js [hunk 3]
```diff
@@ -5,10 +5,10 @@ import {
 } from 'core/constants';
 
 
-export function searchStart({ filters, page, results }) {
+export function searchStart({ errorHandlerId, filters, page, results }) {
```
Comment (kumar303) [inline src/core/actions/search.js]: Since all of these parameters are required, the function should throw an error if any are empty. This is important because action creators are one part of the redux flow that we can't capture and replay. In other words, if there is a bug in some calling code where it's passing an empty parameter to an action creator, it might be hard to track down.

File: src/core/sagas/utils.js [hunk 4]
```diff
@@ -1,9 +1,7 @@
-/* @flow */
```
Comment (kumar303) [inline src/core/sagas/utils.js]: Sorry if this was giving you grief. I think it's important to keep Flow where we have it, especially since the problem I mentioned above (possible missing params in your action creator) are protected nicely by Flow. Here is a patch to put it back for you:

```diff
diff --git a/src/core/sagas/utils.js b/src/core/sagas/utils.js
index 65e66069..f35ab181 100644
--- a/src/core/sagas/utils.js
+++ b/src/core/sagas/utils.js
@@ -1,7 +1,12 @@
+/* @flow */
 import { ErrorHandler } from 'core/errorHandler';
 import defaultLog from 'core/logger';
+import type { ApiStateType } from 'core/reducers/api';
 
-export function createErrorHandler(id, { log = defaultLog } = {}) {
+export function createErrorHandler(
+  id: string,
+  { log = defaultLog }: {| log: typeof defaultLog |} = {}
+): typeof ErrorHandler {
   return new ErrorHandler({
     id,
     // Make sure the dispatch() method can't be used. A saga will yield
@@ -12,11 +17,11 @@ export function createErrorHandler(id, { log = defaultLog } = {}) {
 }
 
 // Convenience function to extract API info.
-export function getApi(state) {
+export function getApi(state: {| api: ApiStateType |}): ApiStateType {
   return state.api;
 }
 
 // Convenience function to extract authentication info.
-export function getAuth(state) {
+export function getAuth(state: Object) {
   return state.auth;
 }
```

File: src/core/sagas/utils.js [hunk 5]
```diff
@@ -14,6 +12,11 @@ export function createErrorHandler(id: string): typeof ErrorHandler {
 }
 
 // Convenience function to extract API info.
-export function getApi(state: {| api: ApiStateType |}): ApiStateType {
+export function getApi(state) {
   return state.api;
 }
+
+// Convenience function to extract authentication info.
+export function getAuth(state) {
```
Comment (kumar303) [inline src/core/sagas/utils.js]: can't we just use one function now?

```js
type StateType = {|
  api: ApiStateType,
  auth: Object,
|};

export function getState(state: StateType) {
  return state;
}
```

File: tests/unit/core/sagas/testUtils.js (hunk 1 above)
Comment (kumar303) [inline tests/unit/core/sagas/testUtils.js]: Thanks for adding test coverage! We really need to get back up to 100% coverage so we don't miss stuff like this.

File: src/amo/sagas/index.js (hunk 2 above)
Comment (tofumatt) [inline src/amo/sagas/index.js]: Not using `all()` causes a deprecation warning from `redux-saga`. See the release notes in: https://github.com/redux-saga/redux-saga/releases/tag/v0.15.0

I can add that link to the comment/commit if you'd like! 😄

File: src/core/actions/search.js (hunk 3 above)
Comment (tofumatt) [inline src/core/actions/search.js]: Fair enough. I think results might not actually be required in a refactor later but I will make them required for now.

File: src/core/sagas/utils.js (hunk 4 above)
Comment (tofumatt) [inline src/core/sagas/utils.js]: I removed it because I find that syntax really head-scratching and it really polluted an otherwise quite straightforward file. I honestly have little idea looking at that function signature what that code does anymore, and I can't imagine a relative ES6 newcomer groking it well.

File: src/core/sagas/utils.js (hunk 5 above)
Comment (tofumatt) [inline src/core/sagas/utils.js]: I guess so. I was thinking it was cleaner not to just extract the entire state in case we want to manipulate certain things in the future, but I guess we would have state extraction helpers elsewhere that weren't saga-specific, so yeah.

File: src/amo/sagas/index.js (hunk 2 above)
Comment (kumar303) [inline src/amo/sagas/index.js]: No need, I believe you! I see in the release notes that it's no different than before; it's just more explicit. Works for me.

File: src/core/sagas/utils.js (hunk 4 above)
Comment (kumar303) [inline src/core/sagas/utils.js]: One thing I find really nice about annotating the `state` object is that you know exactly what you're getting. You can follow `ApiStateType` and find out what all the `state.api.*` properties are that you get access to. This would be hugely informative to an ES6 newcomer who is trying to understand how our app works. In fact, they could be pretty lost without it! Obviously it's not helpful in the case of `state.auth` because we don't have types for that one yet.

Comment (tofumatt) [inline src/core/sagas/utils.js]: Heh, I like the idea but an ES6 newcomer is likely to be intimidated by flow's syntax.

Now we're returning all of the state object in the new `getState` function. I guess I can add `Object` to the rest of the state objects (or omit them as you did below) but that feels even worse: we are half-documenting the keys returned by the helper.

File: src/core/actions/search.js (hunk 3 above)
Comment (tofumatt) [inline src/core/actions/search.js]: Making these required meant a lot of changes, so the PR diff definitely grew.

**Comment by tofumatt** [comment]
Shoot, I think I forgot to add coverage for the required params. I'll do that now.

File: src/core/sagas/utils.js [hunk 6]
```diff
@@ -1,19 +1,17 @@
...
-      'ErrorHandler cannot dispatch from a saga'),
+    dispatch: () => log.error('ErrorHandler cannot dispatch from a saga'),
   });
 }
 
-// Convenience function to extract API info.
-export function getApi(state: {| api: ApiStateType |}): ApiStateType {
-  return state.api;
+// Convenience function to extract state info.
+export function getState(state) {
```
Comment (kumar303) [inline src/core/sagas/utils.js]: I think an annotation of `api: ApiStateType` would be really helpful here, especially since `state` is now completely non-intuitive. You mentioned that it's only half-documented since we don't have an `auth` definition but half is better than nothing. How could someone even look at this function and know what to do with `state`?

Also, I'm fine with not adding `/* @flow */` to new files but removing it from existing ones? You need to provide a solid argument for that and I don't see one here. It's hard enough for us to get decent Flow coverage; what's the point of reducing our coverage? If it's slowing you down on this patch, just let me know and I can fix it up on master once you land it.

File: tests/unit/amo/components/TestCategory.js [hunk 7]
```diff
@@ -124,7 +124,12 @@ describe('Category.mapStateToProps()', () => {
   });
 
   it('passes the search state if the filters and state matches', () => {
-    store.dispatch(searchStart({ filters, results: [] }));
+    store.dispatch(searchStart({
```
Comment (kumar303) [inline tests/unit/amo/components/TestCategory.js]: This should use a helper like `_searchStart()` that sets default argument values so that you aren't repeating yourself with arguments that are not relevant to what you are testing.

File: tests/unit/core/actions/test_search.js [hunk 8]
```diff
@@ -1,15 +1,60 @@
...
   });
 
   it('sets the query and existing results', () => {
-    expect(action.payload).toEqual({ filters: { query: 'foo' }, page: 5, results: [] });
+    expect(action.payload).toEqual(payload);
+  });
+
+  it('throws an error if errorHandlerId is empty', () => {
+    expect(() => {
+      actions.searchStart({
```
Comment (kumar303) [inline tests/unit/core/actions/test_search.js]: These also should all use a helper function that sets default values since those values aren't relevant to each test

File: tests/unit/core/sagas/testSearch.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+  it('clears the error handler', async() => {
+    const entities = sinon.stub();
+    const result = sinon.stub();
+
+    const filters = { query: 'test' };
+
+    mockApi
+      .expects('search')
+      .once()
+      .withArgs({ api: apiState, auth: authState, filters, page: 1 })
```
Comment (kumar303) [inline tests/unit/core/sagas/testSearch.js]: I don't think you need to make an assertion about the calling arguments here because there is already a test for that. This specific test only needs an API response to set up a condition for clearing the error.

File: tests/unit/core/sagas/testSearch.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+      .returns(Promise.resolve({ entities, result }));
+
+    _searchStart({
+      filters,
+      page: 1,
```
Comment (kumar303) [inline tests/unit/core/sagas/testSearch.js]: `page` and `results` look like arguments that are not relevant to this test or the ones below it. They can be moved into `_searchStart()` as defaults so that you don't have to repeat them each time.

File: src/core/sagas/search.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+    yield put(searchLoad({ page, filters, ...response }));
+  } catch (error) {
+    log.warn(`Search results failed to load: ${error}`);
+    yield put(errorHandler.createErrorAction(error));
+  }
+}
+
+export default function* searchSaga() {
+  yield takeEvery(SEARCH_STARTED, fetchSearchResults);
```
Comment (kumar303) [inline src/core/sagas/search.js]: Does this need to process parallel search requests? I don't think we have any page that does that currently. Maybe `takeLatest` is enough so that the last request gets cancelled?

File: src/amo/store.js [hunk 12]
```diff
@@ -19,27 +19,29 @@ import search from 'core/reducers/search';
 import { middleware } from 'core/store';
 
 
+export const AMO_REDUCERS = {
```
Comment (kumar303) [inline src/amo/store.js]: any reason why this is in all caps? I feel like that's a bad habit we've gotten into.

File: tests/unit/core/sagas/testUtils.js [hunk 13]
```diff
@@ -6,30 +6,36 @@ import SagaTester from 'redux-saga-tester';
...
-    store.dispatch(setLang('en-US'));
-
+    const { store } = dispatchSignInActions();
     const state = store.getState();
 
     const sagaTester = new SagaTester({
-      initialState: { api: state.api },
-      reducers: { api: apiReducer },
+      initialState: state,
+      reducers: AMO_REDUCERS,
```
Comment (kumar303) [inline tests/unit/core/sagas/testUtils.js]: what do we gain by passing all of AMO's reducers to this tester? I am struggling to think of a benefit. Every reducer gets execute on every action dispatch so it will introduce slowness. The point of this test is just to make sure the entire state is exposed but you can still prove that with a limited reducer map such as `{ api: apiReducer, auth: authReducer }`.

**Comment by kumar303** [comment]
Please also merge in master and fix the call to `yield select(getApi)` in `amo/sagas/landing`.

File: src/core/sagas/utils.js (hunk 6 above)
Comment (tofumatt) [inline src/core/sagas/utils.js]: I find the syntax for objects with default values (eg `{ log = defaultLog }: {| log: typeof defaultLog |} = {}`) to be totally hostile and it's really frustrating to me. But I think I've made the best case I can against flow already which is that I find the syntax to be actively distracting.

I think `state` is largely understood to be our state object, the result of `store.getState()` from our redux `store` object. That's not intuitive if you haven't used a react/redux app (or maybe even haven't used our app), but I think flow's annotations are even less intuitive than the `state` object in this function.

When we're often declaring the types in a separate file I still need to check somewhere else to see exactly what `ApiStateType` is anyway. I've yet to hit a bug that flow would have caught. That's perhaps a bit chicken and egg because we lack any substantial flow coverage, but to me the syntax fights with me more than the bugs it would catch were we to have coverage.

File: src/core/sagas/search.js (hunk 11 above)
Comment (tofumatt) [inline src/core/sagas/search.js]: Good call. I think I confused calling the `search` API multiple times (which the landing page does) with the saga accepting multiple requests.

File: src/amo/store.js (hunk 12 above)
Comment (tofumatt) [inline src/amo/store.js]: Heh, I feel like it's just how one does constants, so nothing other than convention.

File: tests/unit/core/sagas/testUtils.js (hunk 13 above)
Comment (tofumatt) [inline tests/unit/core/sagas/testUtils.js]: The way the test was written the entire state was being returned so using all the reducers made it work... I'll take a look again and see about checking a few things. I thought this was less fragile than picking specific reducers which might eventually change.

File: tests/unit/core/sagas/testSearch.js (hunk 10 above)
Comment (tofumatt) [inline tests/unit/core/sagas/testSearch.js]: You're right, I hastily added them in after adding the checks in the action creator, but they should be in a helper function. Sorry about that.

Comment (tofumatt) [inline tests/unit/core/sagas/testSearch.js]: They actually may both go away in a future patch too, so best to keep 'em in one place so they're easier to refactor out. 👍

File: tests/unit/core/sagas/testUtils.js (hunk 13 above)
Comment (tofumatt) [inline tests/unit/core/sagas/testUtils.js]: Oh I was silly, yeah there's a fine way to do it with specific reducers. Changed!

**Comment by tofumatt** [comment]
Restored flow and added the helpers to the test files. Also removed adding all the reducers.

File: tests/unit/core/actions/test_search.js [hunk 14]
```diff
@@ -1,15 +1,53 @@
...
+      errorHandlerId: 'Search',
+      filters: { query: 'foo' },
+      page: 1,
+      results: [],
+    });
+  });
+
+  it('throws an error if errorHandlerId is empty', () => {
+    expect(() => {
+      _searchStart({ errorHandlerId: undefined });
```
Comment (kumar303) [inline tests/unit/core/actions/test_search.js]: Huh. I didn't even know `undefined` would override the default attributes in the spread but it totally does.

File: tests/unit/core/sagas/testSearch.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+  it('clears the error handler', async() => {
+    const entities = sinon.stub();
+    const result = sinon.stub();
+
+    const filters = { query: 'test' };
+
+    mockApi
+      .expects('search')
+      .once()
+      .withArgs({ api: apiState, auth: authState, filters, page: 1 })
```
Comment (kumar303) [inline tests/unit/core/sagas/testSearch.js]: I don't think you need to make an assertion about the calling arguments here because there is already a test for that. This specific test only needs an API response to set up a condition for clearing the error.

File: src/core/sagas/utils.js [hunk 16]
```diff
@@ -1,19 +1,24 @@
...
 import { ErrorHandler } from 'core/errorHandler';
-import log from 'core/logger';
+import defaultLog from 'core/logger';
 import type { ApiStateType } from 'core/reducers/api';
 
-export function createErrorHandler(id: string): typeof ErrorHandler {
+
+export function createErrorHandler(
+  id: string,
+  { log = defaultLog }: {| log: typeof defaultLog |} = {}
```
Comment (tofumatt) [inline src/core/sagas/utils.js]: There's nothing readable about that syntax to me. 😢 It looks like information overload and it reads like the annotation isn't an addendum to the code/argument but sandwiched in the middle of the default argument declaration. So I'm reading about what the argument is in the code and "bam!" in the middle is annotation before it reverts to code. I think this is the most frustrating part of flow to me: it isn't an annotation as much as it is an interruption.

File: src/core/sagas/utils.js [hunk 17]
```diff
@@ -1,19 +1,24 @@
...
   });
 }
 
-// Convenience function to extract API info.
-export function getApi(state: {| api: ApiStateType |}): ApiStateType {
-  return state.api;
+// Convenience function to extract state info.
+export function getState(
+  state: {| api: ApiStateType, auth: Object |}
+): Object {
```
Comment (tofumatt) [inline src/core/sagas/utils.js]: I find flow's error outputs to be hard to follow as well, but I see what that's doing.

File: tests/unit/amo/components/TestSearchContextCard.js [hunk 18]
```diff
@@ -36,7 +36,12 @@ describe('SearchContextCard', () => {
 
   it('should render "searching" while loading without query', () => {
     const { store } = dispatchClientMetadata();
-    store.dispatch(searchStart({ filters: {} }));
+    store.dispatch(searchStart({
```
Comment (tofumatt) [inline tests/unit/amo/components/TestSearchContextCard.js]: True! Good catch.

File: tests/unit/core/actions/test_search.js (hunk 14 above)
Comment (tofumatt) [inline tests/unit/core/actions/test_search.js]: Yeah it's kind of weird but I think it explains well enough what's happening.

File: tests/unit/core/sagas/testSearch.js (hunk 15 above)
Comment (tofumatt) [inline tests/unit/core/sagas/testSearch.js]: Oh, true, sorry!

File: src/core/sagas/search.js [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+    const state = yield select(getState);
+
+    const response = yield call(searchApi, {
+      api: state.api,
+      auth: state.auth,
+      filters,
+      page,
+    });
+
+    yield put(searchLoad({ page, filters, ...response }));
```
Comment (kumar303) [inline src/core/sagas/search.js]: Oops, I just noticed the `...response` spread. Can you change this to explicit parameters? I know this is just copied from the old code but we've had too much trouble with spreads. 

Also, `searchLoad()` doesn't take a `page` argument at all. Flow helps to catch these types of problems but until you're using it, please double check all function signatures manually.

File: tests/unit/core/sagas/testSearch.js [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+  it('searches the API for add-ons', async() => {
+    const entities = sinon.stub();
+    const result = sinon.stub();
+
+    const filters = { query: 'test' };
+
+    mockApi
+      .expects('search')
+      .once()
+      .withArgs({ api: apiState, auth: authState, filters, page: 1 })
```
Comment (kumar303) [inline tests/unit/core/sagas/testSearch.js]: You don't need to make an assertion about the calling arguments here because there is already a test for that. This specific test only needs an API response to set up a condition for clearing the error. You can delete `withArgs()` on this line.

File: src/core/sagas/search.js (hunk 19 above)
Comment (tofumatt) [inline src/core/sagas/search.js]: Totally, makes the code so much nicer anyhow!

**Comment by tofumatt** [comment]
Changes made, set for another r?

**Comment by kumar303** [comment]
...if Travis says so

**Comment by kumar303** [comment]
Thanks for doing this in a separate patch. I'm going to use the saga for converting the Category page (https://github.com/mozilla/addons-frontend/issues/2857). I'll land it so I can get started!
